crates/cargo-util/src/process_builder.rs

```diff
@@ -550,14 +547,87 @@ mod imp {
     }
 }
 
+/// Escapes and unescapes an command-line argument for the argfile of rustc.
+#[derive(Debug)]
+pub struct ArgEscape<'a> {
+    arg: Cow<'a, str>,
+}
+
+impl fmt::Display for ArgEscape<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.escape().fmt(f)
+    }
+}
+
+impl<'a> From<&'a str> for ArgEscape<'a> {
+    fn from(arg: &'a str) -> Self {
+        ArgEscape {
+            arg: Cow::Borrowed(arg),
+        }
+    }
+}
+
+impl<'a> ArgEscape<'a> {
+    /// Escapes the argument.
+    ///
+    /// - `\n` -> `\\n`
+    /// - `\r` -> `\\r`
+    /// - `\\` -> `\\\\`
+    pub fn escape(&self) -> Cow<'_, str> {
+        if self.arg.contains(['\n', '\r', '\\']) {
+            let mut s = String::with_capacity(self.arg.len());
+            for c in self.arg.chars() {
+                match c {
+                    '\n' => s.push_str("\\n"),
+                    '\r' => s.push_str("\\r"),
+                    '\\' => s.push_str("\\\\"),
+                    c => s.push(c),
+                }
+            }
+            Cow::Owned(s)
+        } else {
+            Cow::Borrowed(&self.arg)
+        }
+    }
+
+    /// Unescapes the argument.
+    ///
+    /// - `\\n` -> `\n`
+    /// - `\\r` -> `\r`
+    /// - `\\\\` -> `\\`
+    /// - `\\<any>` -> `<any>`
+    pub fn unescape(&self) -> Cow<'_, str> {
+        if self.arg.contains('\\') {
+            let mut s = String::with_capacity(self.arg.len() - 1);
+            let mut chars = self.arg.chars();
+            while let Some(c) = chars.next() {
+                match c {
+                    '\\' => match chars.next() {
+                        Some('n') => s.push('\n'),
+                        Some('r') => s.push('\r'),
+                        Some(c) => s.push(c),
+                        None => {}
+                    },
+                    c => s.push(c),
+                }
+            }
+            Cow::Owned(s)
+        } else {
+            Cow::Borrowed(&self.arg)
+        }
+    }
+}
```

Comment:
I'm always paranoid of doing escaping wrong.  Is there an existing escaping mechanism in the stdlib that would be useful?

Reply:
Yes there is a [`std::ascii::escape_default`](https://doc.rust-lang.org/stable/std/ascii/fn.escape_default.html) but we need an unescape counterpart and it seems not existing. Will remove this whole escaping part as your advice. See https://github.com/rust-lang/cargo/pull/10546#discussion_r846788690